File: 백준/Silver/1874. 스택 수열/스택 수열.java

import java.util.Scanner;
import java.util.Stack;

public class Main {

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);

		int t = sc.nextInt();
		int tt = 1;
		// 8
		int num = 1;
		Stack<Integer> stack = new Stack<>();
		StringBuilder sb = new StringBuilder();
		while (tt <= t) {
//			System.out.println(stack);
			int ele = sc.nextInt();

			if (stack.isEmpty() && ele < num) {
				System.out.println("NO");
				return;
			}

			if (ele < num && ele > stack.peek()) {
				System.out.println("NO");
				return;
			}

			while (ele >= num) {
				stack.add(num++);
				sb.append("+\n");
			}

			while (!stack.isEmpty() && ele <= stack.peek()) {
				stack.pop();
				sb.append("-\n");
			}

			tt++;
		}
		System.out.println(sb);

	}

}
//
//5
//1
//5
//3
//4
//2

// 1 2 5 55555555555555555

File: 백준/Silver/1874. 스택 수열/스택 수열.py
n = int(input())
stack = []
idx = 1
ans = ""
first = int(input())
while (idx <= first):
    stack.append(idx)
    idx += 1
    ans += "+\n"
stack.pop()
ans += "-\n"
for i in range(n-1):
    num = int(input())
    while (idx <= num):
        stack.append(idx)
        idx += 1
        ans+="+\n"
    while(stack and stack[-1]!=num):
        stack.pop()
        ans+="-\n"
    if(not stack):
        ans = "NO"
        break
    elif(stack[-1]==num):
        stack.pop()
        ans+="-\n"
print(ans)